projects/project3/Mining Association rules from Semantic data using Spark/src/main/scala/net/sansa_stack/template/spark/rdf/RulesMining.scala

```diff
@@ -0,0 +1,438 @@
+import scala.collection.mutable.ListBuffer
+import scala.collection.mutable.HashMap
+import scala.collection.mutable.HashSet
+import java.util.ArrayList
+import scala.util.control.Breaks._
+import org.apache.spark.mllib.fpm.FPGrowth.FreqItemset
+import org.apache.spark.mllib.fpm.AssociationRules
+
+object RulesMining {
+
+  var featureMap: HashMap[String, String] = new HashMap[String, String] //map of nodes with feature
+  var compTriples = List.newBuilder[Triple[String, List[String], String]] //list of composition triples
+  var instanceTransactions: ListBuffer[HashSet[String]] = new ListBuffer[HashSet[String]] //list of instance transactions
+  var targetConceptNodesList: ListBuffer[String] = new ListBuffer[String] //nodes in graph with feature as target concept
+  var targetConceptNeighboursList: ListBuffer[Integer] = new ListBuffer[Integer] //immediate nodes of target concept in the graph
+  var targetConceptVertexId: Int = 0 //vertexId of targetConceptNodes in graph
+  var targetConceptNeighboursId: Int = 0 //vertexId of neighbours of targetConcept
+  val query: String = "Patient,Report,{Disease,Drug,damageIndex}" //query pattern "targetConcept,context,{features+}"
+  type VertexId = Long // vertex id for every node vertex in graph
+  var testList: List[String] = List.empty[String] //list of modified composition triples for item generation
+  var finalListReport: List[String] = List.empty[String] //list of item transactions
+
+  System.setProperty("hadoop.home.dir", "G:/Bonn/Sem2/BigDataLab/"); //File location should be specified
+  def main(args: Array[String]) {
+    parser.parse(args, Config()) match {
+      case Some(config) =>
+        run(config.in)
+      case None =>
+        println(parser.usage)
+    }
+  }
+
+  def run(input: String): Unit = {
+
+    val spark = SparkSession.builder
+      .appName(s"RulesMining $input")
+      .master("local[*]")
+      .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
+      .getOrCreate()
+
+    println("======================================")
+    println("|            Reading Input           |")
+    println("======================================")
+
+    //Reading the nTriples file
+    val triplesRDD = NTripleReader.load(spark, JavaURI.create(input))
+    val tripleRDD = triplesRDD.map(triple => (triple.getSubject.toString, triple.getPredicate.toString, triple.getObject.toString))
+    updateFeatures()
+    val turtleSubjectObject = tripleRDD.map { x => (x._1, x._3) }
+
+    val indexVertexID = (tripleRDD.map(_._1) union tripleRDD.map(_._3)).distinct().zipWithIndex()
+    val vertices: RDD[(VertexId, String)] = indexVertexID.map(f => (f._2, f._1))
+
+    val tuples = tripleRDD.keyBy(_._1).join(indexVertexID).map(
+      {
+        case (k, (Triple(s, p, o), si)) => (o, (si, p))
+      })
+
+    val edges: RDD[Edge[String]] = tuples.join(indexVertexID).map({
+      case (k, ((si, p), oi)) => Edge(si, oi, p)
+    })
+
+    val graph = Graph(vertices, edges)

```

Comment:
You could use our own [`asGraph`](https://github.com/SANSA-Stack/SANSA-Examples/blob/develop/sansa-examples-spark/src/main/scala/net/sansa_stack/examples/spark/rdf/PageRank.scala#L40) method from SANSA in case your graph transformation of triples is a standard one which we use as well.

Reply:
Please, consider this option when you build the graph.